### PR TITLE
Sharded DDP Docs

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Fixes # (issue)

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/multi_gpu.rst [hunk 1]
```diff
@@ -429,6 +429,33 @@ You can then call your scripts anywhere
     cd /project/src
     python some_file.py --accelerator 'ddp' --gpus 8
 
+Sharded DDP
```
Comment (edenlightning) [inline docs/source/multi_gpu.rst:432]:
```suggestion
Sharded DDP [BETA]
```

File: docs/source/multi_gpu.rst [hunk 2]
```diff
@@ -429,6 +429,33 @@ You can then call your scripts anywhere
...
+^^^^^^^^^^^
+Lightning integration of `DeepSpeed ZeRO <https://arxiv.org/abs/1910.02054>`_ and
+`ZeRO-2 <https://www.microsoft.com/en-us/research/blog/zero-2-deepspeed-shattering-barriers-of-deep-learning-speed-scale/>`_
+provided by `Fairscale <https://github.com/facebookresearch/fairscale/tree/master/fairscale>`_. Sharded DDP is similar to normal DDP, except optimizer state and gradients are sharded across your GPUs.
+This means the memory overhead per GPU is less, as each GPU only has to maintain a section of your optimizer state and gradients.
```
Comment (edenlightning) [inline docs/source/multi_gpu.rst:437]:
```suggestion
This means the memory overhead per GPU is lower, as each GPU only has to maintain a section of your optimizer state and gradients.
```

File: docs/source/multi_gpu.rst [hunk 3]
```diff
@@ -429,6 +429,33 @@ You can then call your scripts anywhere
...
+
+The benefits are variable by model, but we've recorded up to a 25% memory reduction per GPU. Because of extremely efficient communication,
+these benefits in multi-GPU setups are almost free and throughput scales well with multi-node setups.
+
+It is highly recommended to use Sharded DDP in multi-GPU environments where memory is limited or where training larger models are beneficial.
+
+Before running, install Fairscale using the command below or install all extras using ``pip install pytorch-lightning["extra"]``.
```
Comment (edenlightning) [inline docs/source/multi_gpu.rst:444]:
```suggestion
To use shared DDP, you need to first install Fairscale using the command below or install all extras using ``pip install pytorch-lightning["extra"]``.
```

File: docs/source/multi_gpu.rst [hunk 4]
```diff
@@ -429,7 +429,6 @@ You can then call your scripts anywhere
     cd /project/src
     python some_file.py --accelerator 'ddp' --gpus 8
 
-
```
Comment (edenlightning) [inline docs/source/multi_gpu.rst:432]: rmove?

File: docs/source/multi_gpu.rst [hunk 5]
```diff
@@ -640,16 +639,59 @@ The reason is that the full batch is visible to all GPUs on the node when using
...
 
-PytorchElastic
+Model Parallelism [BETA]
+------------------------
+
+Model Parallelism tackles training large models on distributed systems, by modifying distributed communications and memory management of the model.
+Unlike data parallelism, the model is partitioned in various ways across the GPUs, in most cases to reduce the memory overhead.
+This is useful when dealing with large Transformer based models, or in environments where GPU memory is limited.
+
+Lightning currently offers these methods to leverage model parallelism.
```
Comment (edenlightning) [inline docs/source/multi_gpu.rst:649]:
```suggestion
Lightning currently offers the following methods to leverage model parallelism:
```

File: docs/source/multi_gpu.rst [hunk 6]
```diff
@@ -640,16 +639,59 @@ The reason is that the full batch is visible to all GPUs on the node when using
 
 ----------
 
-PytorchElastic
+Model Parallelism [BETA]
```
Comment (edenlightning) [inline docs/source/multi_gpu.rst:642]: Maybe above batch size?

File: docs/source/multi_gpu.rst [hunk 7]
```diff
@@ -640,16 +639,59 @@ The reason is that the full batch is visible to all GPUs on the node when using
...
+Lightning integration of optimizer sharded training provided by `Fairscale <https://github.com/facebookresearch/fairscale>`_.
+The technique can be found within `DeepSpeed ZeRO <https://arxiv.org/abs/1910.02054>`_ and
+`ZeRO-2 <https://www.microsoft.com/en-us/research/blog/zero-2-deepspeed-shattering-barriers-of-deep-learning-speed-scale/>`_,
+however the implementation is built from the ground up to be pytorch compatible and standalone.
+
+Optimizer Sharded Training still utilizes Data Parallel Training under the hood, except optimizer state and gradients are sharded across GPUs.
+This means the memory overhead per GPU is lower, as each GPU only has to maintain a section of your optimizer state and gradients.
+
+The benefits are variable by model, but we've recorded up to a 25% memory reduction per GPU. Because of extremely efficient communication,
+these benefits in multi-GPU setups are almost free and throughput scales well with multi-node setups.
```
Comment (edenlightning) [inline docs/source/multi_gpu.rst:664]: Are there any scenarios where this is not recommended?

File: docs/source/performance.rst [hunk 8]
```diff
@@ -114,3 +114,30 @@ However, know that 16-bit and multi-processing (any DDP) can have issues. Here a
...
+-------------------------------------------------------
+
+Sharded DDP is a lightning integration of `DeepSpeed ZeRO <https://arxiv.org/abs/1910.02054>`_ and
+`ZeRO-2 <https://www.microsoft.com/en-us/research/blog/zero-2-deepspeed-shattering-barriers-of-deep-learning-speed-scale/>`_
+provided by `Fairscale <https://github.com/facebookresearch/fairscale>`_.
+
+When training on multiple GPUs sharded DDP can assist to increase memory efficiency substantially, and in some cases performance on multi-node is better than traditional DDP.
+This is due to efficient communication and parallelization under the hood.
+
+Before running, install Fairscale using the command below or install all extras using ``pip install pytorch-lightning["extra"]``.
```
Comment (edenlightning) [inline docs/source/performance.rst:130]:
```suggestion
To use Optimizer Sharded Training, you need to first install Fairscale using the command below or install all extras using ``pip install pytorch-lightning["extra"]``.
```

File: docs/source/multi_gpu.rst [hunk 9]
```diff
@@ -598,6 +598,53 @@ If you need your own way to init PyTorch DDP you can override :meth:`pytorch_lig
...
+- Optimizer Sharded Training (partitioning your gradients and optimizer state across multiple GPUs, for reduced memory overhead)
+
+Optimizer Sharded Training
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+Lightning integration of optimizer sharded training provided by `Fairscale <https://github.com/facebookresearch/fairscale>`_.
+The technique can be found within `DeepSpeed ZeRO <https://arxiv.org/abs/1910.02054>`_ and
+`ZeRO-2 <https://www.microsoft.com/en-us/research/blog/zero-2-deepspeed-shattering-barriers-of-deep-learning-speed-scale/>`_,
+however the implementation is built from the ground up to be pytorch compatible and standalone.
+
+Optimizer Sharded Training still utilizes Data Parallel Training under the hood, except optimizer state and gradients are sharded across GPUs.
```
Comment (edenlightning) [inline docs/source/multi_gpu.rst:621]:
```suggestion
Optimizer Sharded Training still utilizes Data Parallel Training under the hood, except the optimizer state and gradients which are sharded across GPUs.
```

File: docs/source/multi_gpu.rst [hunk 10]
```diff
@@ -598,6 +598,53 @@ If you need your own way to init PyTorch DDP you can override :meth:`pytorch_lig
...
+This means the memory overhead per GPU is lower, as each GPU only has to maintain a section of your optimizer state and gradients.
```
Comment (edenlightning) [inline docs/source/multi_gpu.rst:622]:
```suggestion
This means the memory overhead per GPU is lower, as each GPU only has to maintain a partition of your optimizer state and gradients.
```

File: docs/source/multi_gpu.rst [hunk 11]
```diff
@@ -598,6 +598,53 @@ If you need your own way to init PyTorch DDP you can override :meth:`pytorch_lig
...
+
+The benefits are variable by model, but we've recorded up to a 63% memory reduction per GPU allowing us to double our model sizes. Because of extremely efficient communication,
```
Comment (edenlightning) [inline docs/source/multi_gpu.rst:624]: maybe vary? not sure

**Review by edenlightning — APPROVED** [review]

File: docs/source/performance.rst [hunk 12]
```diff
@@ -114,3 +114,30 @@ However, know that 16-bit and multi-processing (any DDP) can have issues. Here a
...
+.. code-block:: bash
+
+    pip install https://github.com/facebookresearch/fairscale/archive/master.zip
+
+.. code-block:: python
+
+    # train using Sharded DDP
+    trainer = Trainer(accelerator='ddp', plugins='ddp_sharded')
+
+Sharded DDP can work across all DDP variants by adding the additional ``--plugins ddp_sharded`` flag.
```
Comment (Borda) [inline docs/source/performance.rst:141]: do we need it twice,m it the same as above, rather make cross-reference?

**Review by teddykoker — APPROVED** [review]

**Review by Borda — APPROVED** [review]
lgtm

**Review by ananyahjha93 — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: docs/source/multi_gpu.rst [hunk 13]
```diff
@@ -598,6 +598,55 @@ If you need your own way to init PyTorch DDP you can override :meth:`pytorch_lig
...
+
+
+.. code-block:: python
+
+    # train using Sharded DDP
+    trainer = Trainer(accelerator='ddp', plugins='ddp_sharded')
+
+Optimizer Sharded Training can work across all DDP variants by adding the additional ``--plugins ddp_sharded`` flag.
+
+Internally we re-initialize optimizers, sharding the optimizers across your machines and processes. We handle all communication using PyTorch distributed, so no code changes are required.
```
Comment (tchaton) [inline docs/source/multi_gpu.rst:647]:
```suggestion
Internally we re-initialize your optimizers and shard them across your machines and processes. We handle all communication using PyTorch distributed, so no code changes are required.
```
